File: C/Queue/QueueUsingLL.c
#include<stdio.h>
#include<stdlib.h>

struct Node {
    int data;
    struct Node *next;
};
struct Node *temp,*rear=NULL,*front=NULL,*newnode;


void N_Queue(){
    struct Node *newnode; 
    newnode = (struct Node*)malloc(sizeof(struct Node));
    printf("Enter data You Want to Push: ");
    scanf("%d", &newnode->data);
    if(rear==NULL&& front==NULL){
        newnode->next=NULL;
        front=newnode;
        rear=newnode;
        rear->next=NULL;        
        }
        else{
            rear=newnode;
            newnode->next=NULL;
            rear=newnode;      
            }
    }
    
 void D_Queue(){
     int x;
     if(rear==NULL&& front==NULL){
         printf("Queue is Empty!!!");
         }
         else{
             temp=front;
             front=front->next;
             x=temp->data;
             free(temp);
             printf("Poped Element %d",x);
             }
     
     }
     void Display() {
    struct Node *temp;
    temp = front;
    printf("[ ");
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("]");
}
int main() {
    int ch;
    do {
        printf("\n\nQUEUE Operation:\n1. N Queue\n2. D Queue\n3. DISPLAY\n4. EXIT\n");
        printf("Enter your choice: ");
        scanf("%d", &ch);
        switch (ch) {
            case 1:
                N_Queue();
                break;
            case 2:
                 D_Queue();
                break;
            case 3:
                Display();
                break;
            case 4:
                printf("Operation Ended !!!\n");
                break;
            default:
                printf("Invalid Choice\n");
                break;
        }
    } while (ch != 4);
    return 0;
    }

File: 01-Java/Data Structure/Queue/QueueUsingLL.java
//queue using Linked List

public class QueueB {

   static class Node {

       int data;

       Node next;

       Node(int data) {

           this.data = data;

           next = null;

       }

   }


   static class Queue {

       static Node head = null;

       static Node tail = null;


       public static boolean isEmpty() {

           return head == null && tail == null;

       }


       public static void add(int data) {

           Node newNode = new Node(data);

           if(isEmpty()) {

               tail = head = newNode;

           } else {

               tail.next = newNode;

               tail = newNode;

           }

       }


       public static int remove() {

           if(isEmpty()) {

               System.out.println("empty queue");

               return -1;

           }

           int front = head.data;

           //single node

           if(head == tail) {

               tail = null;

           }

           head = head.next;

           return front;

       }


       public static int peek() {

           if(isEmpty()) {

               System.out.println("empty queue");

               return -1;

           }

         

           return head.data;

       }

   }

   public static void main(String args[]) {

       Queue q = new Queue();

       q.add(1);

       q.add(2);

       q.add(3);

       q.add(4);

       q.add(5);


       while(!q.isEmpty()) {

           System.out.println(q.peek());

           q.remove();

       }

   }

}

